core/src/player.rs

```diff
@@ -1768,6 +1654,305 @@ impl Player {
+                mouse_pos: (Twips::ZERO, Twips::ZERO),
+                mouse_cursor: MouseCursor::Arrow,
+                mouse_cursor_needs_check: false,
+
+                // Misc. state
+                rng: SmallRng::seed_from_u64(chrono::Utc::now().timestamp_millis() as u64),
+                system: SystemProperties::default(),
+                transform_stack: TransformStack::new(),
+                instance_counter: 0,
+                player_version: NEWEST_PLAYER_VERSION,
+                is_playing: self.autoplay,
+                needs_render: true,
+                warn_on_unsupported_content: self.warn_on_unsupported_content,
+                self_reference: self_ref.clone(),
+
+                // GC data
+                gc_arena: GcArena::new(ArenaParameters::default(), |gc_context| {
+                    GcRoot(GcCell::allocate(
+                        gc_context,
+                        GcRootData {
+                            audio_manager: AudioManager::new(),
+                            action_queue: ActionQueue::new(),
+                            avm1: Avm1::new(gc_context, NEWEST_PLAYER_VERSION),
+                            avm2: Avm2::new(gc_context),
+                            current_context_menu: None,
+                            drag_object: None,
+                            external_interface: ExternalInterface::new(),
+                            focus_tracker: FocusTracker::new(gc_context),
+                            library: Library::empty(),
+                            load_manager: LoadManager::new(),
+                            mouse_hovered_object: None,
+                            mouse_pressed_object: None,
+                            shared_objects: HashMap::new(),
+                            stage: Stage::empty(
+                                gc_context,
+                                self.viewport_width,
+                                self.viewport_height,
+                            ),
+                            timers: Timers::new(),
+                            unbound_text_fields: Vec::new(),
+                        },
+                    ))
+                }),
+            })
+        });
+
+        // Finalize configuration and load the movie.
+        let mut player_lock = player.lock().unwrap();
+        player_lock.mutate_with_update_context(|context| {
+            // Instantiate an empty root before the main movie loads.
+            let fake_root = MovieClip::from_movie(context.gc_context, fake_movie);
+            fake_root.post_instantiation(context, None, Instantiator::Movie, false);
+            context.stage.replace_at_depth(context, fake_root.into(), 0);
+
+            let result = Avm2::load_player_globals(context);
+
+            let stage = context.stage;
+            stage.post_instantiation(context, None, Instantiator::Movie, false);
+            stage.build_matrices(context);
+
+            result
+        })?;
```

Comment:
I think this is the only failable place in `PlayerBuilder::build`. Maybe we can avoid it to make the whole function unfailable?

Reply:
Hmm, `Avm2::load_player_globals` is what returns a result, and this should basically "never fail" otherwise we really screwed up anyway -- so maybe we can just `unwrap`?

Reply:
Agree.